src/network/cni/namespace.rs

```diff
@@ -0,0 +1,85 @@
+//! Network namespace helpers and structures.
+
+use anyhow::{Context, Error, Result};
+use getset::Getters;
+use log::trace;
+use nix::{
+    sched::{setns, CloneFlags},
+    unistd::gettid,
+};
+use std::{
+    future::Future,
+    os::unix::io::AsRawFd,
+    path::{Path, PathBuf},
+    process,
+};
+use tokio::fs::File;
+
+#[derive(Debug, Getters)]
+/// A basic network namespace abstraction.
+pub struct Namespace {
+    #[get]
+    file: File,
+}
+
+impl Namespace {
+    /// Create a new namespace.
+    pub async fn new<P>(path: P) -> Result<Self>
+    where
+        P: AsRef<Path>,
+    {
+        Ok(Self {
+            file: File::open(&path).await.context("open namespace file")?,
+        })
+    }
+
+    /// Run a closure inside the namespace
+    pub async fn run<F>(&self, fun: F) -> Result<()>
+    where
+        F: Future<Output = Result<()>> + Send + 'static,
+    {
+        trace!("Using file as network namespace: {:?}", self.file());
+        let fd = self.file().as_raw_fd();
+        tokio::spawn(async move {
+            setns(fd, CloneFlags::CLONE_NEWNET).context("switch to network namespace")?;
+            trace!(
+                "Switched to network namespace: {}",
+                Self::current_thread_namespace_path().display()
+            );
+
+            fun.await.context("run namespace future")?;
+
+            let ok: std::result::Result<(), Error> = Ok(());
+            ok
+        })
```

Comment:
I don't think this is correct -- you are switching the network namespace of a thread inside the tokio threadpool and then running a closure in it but while the closure is not ready you may end up running other code in the thread (not to mention once this closure exits, the thread is still in the different network namespace). For an example what might go wrong if two separate `run` calls happen at the same time and they're scheduled (interleaved) on the same thread in the tokio threadpool?

I would suggest that you want to either spawn a thread explicitly with `std::thread::spawn` (so that you know nothing else runs in that thread, and it dies after the closure is done executing) or you create a thread-pool each time (but it's probably simpler to just use `std::thread::spawn` -- though that means you'll have to handle the async stuff somewhat manually).

Reply:
I think that [`async_std::spawn`](https://docs.rs/async-std/1.6.5/async_std/task/fn.spawn.html) might do what you want but I'm not entirely sure.

Reply:
Another option is to spawn a process using `tokio::process`.

Reply:
Okay, I think I found a proper solution. First I patched server.rs to setup/teardown a network directly on server init:

```patch
diff --git a/src/server.rs b/src/server.rs
index 3abd230..ba32743 100644
--- a/src/server.rs
+++ b/src/server.rs
@@ -45,7 +45,22 @@ impl Server {
         let storage = DefaultKeyValueStorage::open(&self.config.storage_path())?;
         let cri_service = CRIService::new(storage.clone());
 
-        let network = self.initialize_network().await.context("init network")?;
+        let mut network = self.initialize_network().await.context("init network")?;
+
+        use crate::sandbox::{LinuxNamespaces, SandboxDataBuilder};
+        let s = SandboxDataBuilder::default()
+            .id("some_id_21")
+            .name("name")
+            .namespace("namespace")
+            .attempt(1u32)
+            .linux_namespaces(LinuxNamespaces::NET)
+            .hostname("hostname")
+            .log_directory("log_directory")
+            .network_namespace_path("/var/run/netns/test")
+            .annotations(std::collections::HashMap::new())
+            .build()?;
+        network.start(&s).await?;
+        network.stop(&s).await?;
 
         // Build a new socket from the config
         let mut uds = self.unix_domain_listener().await?;
```

Then created the network namespace:
```shell
> sudo ip netns add test
```

Now run the server:
```
> cargo build && sudo ./target/debug/criserver -l trace
```

And verify the logs
```
[2020-10-07T09:32:02.171Z INFO  cri::network::cni src/network/cni/mod.rs:473] Starting CNI network for sandbox some_id_21
[2020-10-07T09:32:02.171Z TRACE cri::network::cni::namespace src/network/cni/namespace.rs:42] Using file as network namespace: tokio::fs::File { std: File { fd: 13, path: "/run/netns/test", read: true, write: false } }
[2020-10-07T09:32:02.171Z TRACE cri::network::cni::namespace src/network/cni/namespace.rs:45] Current thread network namespace: /proc/13859/task/13859/ns/net
[2020-10-07T09:32:02.171Z TRACE cri::network::cni::namespace src/network/cni/namespace.rs:52] Switched to network namespace: /proc/13859/task/13914/ns/net
[2020-10-07T09:32:02.171Z TRACE cri::network::cni src/network/cni/mod.rs:487] Running netlink in network namespace: /proc/13859/task/13914/ns/net
[2020-10-07T09:32:02.171Z DEBUG cri::network::cni::netlink src/network/cni/netlink.rs:35] Creating new netlink connection
[2020-10-07T09:32:02.172Z TRACE cri::network::cni::netlink src/network/cni/netlink.rs:64] Got link by name lo: LinkHeader { interface_family: 0, index: 1, link_layer_type: 772, flags: 8, change_mask: 0 }
[2020-10-07T09:32:02.172Z TRACE cri::network::cni::netlink src/network/cni/netlink.rs:82] Setting link lo up
[2020-10-07T09:32:02.173Z TRACE cri::network::cni::plugin src/network/cni/plugin.rs:56] Using plugin binary /run/current-system/sw/bin/bridge
[2020-10-07T09:32:02.173Z TRACE cri::network::cni::plugin src/network/cni/plugin.rs:140] Using CNI args Args { command: "ADD", container_id: "some_id_21", network_namespace: "/var/run/netns/test", plugin_args: [], interface_name: "eth0", path: "/run/current-system/sw/bin" }
[2020-10-07T09:32:02.173Z TRACE cri::network::cni::exec src/network/cni/exec.rs:120] Using CNI env: {"CNI_CONTAINERID": "some_id_21", "CNI_NETNS": "/var/run/netns/test", "CNI_COMMAND": "ADD", "CNI_ARGS": "", "CNI_IFNAME": "eth0", "CNI_PATH": "/run/current-system/sw/bin"}
[2020-10-07T09:32:04.225Z TRACE cri::network::cni::plugin src/network/cni/plugin.rs:150] Got CNI ouput CNIResult { cni_version: "0.3.1", interfaces: [NetworkInterface { name: "cni0", mac: Some("42:9c:c2:08:a5:53"), sandbox: None }, NetworkInterface { name: "veth913b81f3", mac: Some("0a:c7:5c:e4:c0:6c"), sandbox: None }, NetworkInterface { name: "eth0", mac: Some("52:ec:82:58:31:d6"), sandbox: Some("/var/run/netns/test") }], ips: [IP { version: "4", interface: 2, address: V4(Ipv4Network { addr: 10.85.0.23, prefix: 16 }), gateway: V4(10.85.0.1) }, IP { version: "6", interface: 2, address: V6(Ipv6Network { addr: 1100:200::c5, prefix: 24 }), gateway: V6(1100:200::1) }], routes: [Route { dst: V4(Ipv4Network { addr: 0.0.0.0, prefix: 0 }), gw: None }, Route { dst: V6(Ipv6Network { addr: 1100:200::1, prefix: 24 }), gw: None }], dns: Some(DNS { nameservers: None, domain: None, search: None, options: None }) }
[2020-10-07T09:32:04.226Z DEBUG cri::network::cni src/network/cni/mod.rs:516] Started CNI network for sandbox some_id_21
[2020-10-07T09:32:04.226Z INFO  cri::network::cni src/network/cni/mod.rs:522] Stopping CNI network for sandbox some_id_21
[2020-10-07T09:32:04.226Z TRACE cri::network::cni::namespace src/network/cni/namespace.rs:42] Using file as network namespace: tokio::fs::File { std: File { fd: 13, path: "/run/netns/test", read: true, write: false } }
[2020-10-07T09:32:04.227Z TRACE cri::network::cni::namespace src/network/cni/namespace.rs:45] Current thread network namespace: /proc/13859/task/13859/ns/net
[2020-10-07T09:32:04.227Z TRACE cri::network::cni::namespace src/network/cni/namespace.rs:52] Switched to network namespace: /proc/13859/task/13914/ns/net
[2020-10-07T09:32:04.227Z TRACE cri::network::cni src/network/cni/mod.rs:530] Running netlink in network namespace: /proc/13859/task/13914/ns/net
[2020-10-07T09:32:04.227Z DEBUG cri::network::cni::netlink src/network/cni/netlink.rs:35] Creating new netlink connection
[2020-10-07T09:32:04.229Z TRACE cri::network::cni::netlink src/network/cni/netlink.rs:64] Got link by name lo: LinkHeader { interface_family: 0, index: 1, link_layer_type: 772, flags: 65609, change_mask: 0 }
[2020-10-07T09:32:04.229Z TRACE cri::network::cni::netlink src/network/cni/netlink.rs:70] Setting link lo down
[2020-10-07T09:32:04.231Z TRACE cri::network::cni::plugin src/network/cni/plugin.rs:56] Using plugin binary /run/current-system/sw/bin/bridge
[2020-10-07T09:32:04.231Z TRACE cri::network::cni::plugin src/network/cni/plugin.rs:140] Using CNI args Args { command: "DEL", container_id: "some_id_21", network_namespace: "/var/run/netns/test", plugin_args: [], interface_name: "eth0", path: "/run/current-system/sw/bin" }
[2020-10-07T09:32:04.232Z TRACE cri::network::cni::exec src/network/cni/exec.rs:120] Using CNI env: {"CNI_NETNS": "/var/run/netns/test", "CNI_COMMAND": "DEL", "CNI_CONTAINERID": "some_id_21", "CNI_ARGS": "", "CNI_PATH": "/run/current-system/sw/bin", "CNI_IFNAME": "eth0"}
[2020-10-07T09:32:04.336Z DEBUG cri::network::cni src/network/cni/mod.rs:559] Stopped CNI network for sandbox some_id_21
```

Note that the path to the network namespace (`/proc/13859/task/13914/ns/net`) inside the closure and in the newly spawned thread is the same, so I guess we're safe now.

Reply:
Note that we have to switch back to the host network namespace because the threads will be re-used by tokio. I'll implement that as well.